pants

```diff
@@ -194,23 +198,44 @@ function determine_python_exe {
   die "Invalid Python interpreter version for ${python_exe}. ${requirement_str}"
 }
 
+function compute_sha256 {
+  local python="$1"
+  local path="$2"
+
+  "$python" <<EOF
+import hashlib
+
+hasher = hashlib.sha256()
+with open('${path}', 'rb') as fp:
+    buf = fp.read()
+    hasher.update(buf)
+print(hasher.hexdigest())
+EOF
+}
+
 # TODO(John Sirois): GC race loser tmp dirs leftover from bootstrap_XXX
 # functions.  Any tmp dir w/o a symlink pointing to it can go.
 
-function bootstrap_venv {
-  if [[ ! -d "${PANTS_BOOTSTRAP}/${VENV_PACKAGE}" ]]; then
+function bootstrap_virtualenv {
+  python="$1"
+  if [[ ! -d "${PANTS_BOOTSTRAP}/${VIRTUALENV_ZIPAPP}" ]]; then
     (
+      green "Downloading the virtualenv zipapp"
       mkdir -p "${PANTS_BOOTSTRAP}"
       local staging_dir
       staging_dir=$(tempdir "${PANTS_BOOTSTRAP}")
       cd "${staging_dir}"
-      curl -LO "https://pypi.io/packages/source/v/virtualenv/${VENV_TARBALL}"
-      tar -xzf "${VENV_TARBALL}"
-      ln -s "${staging_dir}/${VENV_PACKAGE}" "${staging_dir}/latest"
-      mv "${staging_dir}/latest" "${PANTS_BOOTSTRAP}/${VENV_PACKAGE}"
-    ) 1>&2
+      curl -LO "${VIRTUALENV_ZIPAPP_URL}"
+      fingerprint="$(compute_sha256 "${python}" "${VIRTUALENV_ZIPAPP}")"
+      if [[ "${VIRTUALENV_ZIPAPP_EXPECTED_SHA256}" != "${fingerprint}" ]]; then
+        die "SHA256 of ${staging_dir}/${VIRTUALENV_ZIPAPP} is not as expected. Aborting."
```

Comment:
Maybe explicitly reference `VIRTUALENV_ZIPAPP_URL`? As a user, I'm not sure I would understand how to fix this issue.